### PR TITLE
Refactor VertexAI and add code-gecko

### PR DESCRIPTION
**Pull request by Lash-L**
## Description

- combine all of vertexai under one provider to allow for easier changing between models and adding new models
- Add code-gecko
- Add ability to add vertex models in ui (currently just the gemini ones)

## Checklist

- [x] The base branch of this PR is `dev`, rather than `main`
- [x] The relevant docs, if any, have been updated or created

## Screenshots

[ For visual changes, include screenshots. ]

## Testing

[ For new or modified features, provide step-by-step testing instructions to validate the intended behavior of the change. ]

### CODE REVIEW

**Comment by Lash-L** [comment]
@reosablo - I don't have access to the mistral or anthropic models on my vertexai instance. If you would be able to test them after this refactor, I would appreciate the peace of mind!

File: gui/src/pages/AddNewModel/configs/providers.ts [hunk 1]
```diff
@@ -626,4 +626,36 @@ To get started, [register](https://dataplatform.cloud.ibm.com/registration/stepo
     ],
     apiKeyUrl: "https://cloud.cerebras.ai/",
   },
+  vertexai: {
+    title: "VertexAI",
+    provider: "vertexai",
+    description: "Use supported vertex AI models",
```
Comment (Patrick-Erichsen) [inline gui/src/pages/AddNewModel/configs/providers.ts]:
```suggestion
    description: "Use supported Vertex AI models",
```

File: gui/src/pages/AddNewModel/configs/providers.ts [hunk 2]
```diff
@@ -626,4 +626,36 @@ To get started, [register](https://dataplatform.cloud.ibm.com/registration/stepo
...
+    longDescription:
+      "Use the supported vertex AI models - see [here](https://cloud.google.com/docs/authentication/provide-credentials-adc) to authenticate",
```
Comment (Patrick-Erichsen) [inline gui/src/pages/AddNewModel/configs/providers.ts]:
```suggestion
      "Use the supported Vertex AI models - see [here](https://cloud.google.com/docs/authentication/provide-credentials-adc) to authenticate",
```

File: gui/src/pages/AddNewModel/configs/providers.ts [hunk 3]
```diff
@@ -626,4 +626,36 @@ To get started, [register](https://dataplatform.cloud.ibm.com/registration/stepo
...
+      models.VertexGemini15Pro,
+      models.VertexGemini15Flash,
+
+    ],
+    collectInputFor: [
+      {
+        inputType: "project",
+        key: "projectId",
+        label: "Project Id",
+        placeholder: "Enter your Vertex AI project id",
```
Comment (Patrick-Erichsen) [inline gui/src/pages/AddNewModel/configs/providers.ts]:
```suggestion
        placeholder: "Enter your Vertex AI project Id",
```

**Review by Patrick-Erichsen — COMMENTED** [review]
This is much cleaner, thanks for all the work to consolidate here!

I think this would be good to merge as is, but it would be ideal if we could reuse the provider logic from the existing Gemini, Anthropic, and Mistral classes instead of copying them over here.

I know that the existing Vertex AI classes already we're doing this copy/paste though so regardless this PR is a big improvement 👍 

cc @sestinj

**Comment by Lash-L** [comment]
@Patrick-Erichsen i can see what I can do! I agree with you 100%. Ill have some time to work on it Saturday

**Comment by Patrick-Erichsen** [comment]
Cool, that would be huge. Again, I think this is ready to merge as is, but it would be quite nice 🙂

File: docs/docs/customize/model-providers/top-level/vertexai.md [hunk 4]
```diff
@@ -11,14 +11,14 @@ You need to enable the [Vertex AI API](https://console.cloud.google.com/marketpl
 
 ## Chat model
 
-We recommend configuring **Mistral Large** as your chat model.
+We recommend configuring **Mistral Large**, **Gemini 1.5 Pro**, or **Claude 3.5 Sonnet** as your chat model.
```
Comment (Patrick-Erichsen) [inline docs/docs/customize/model-providers/top-level/vertexai.md]: @TyDunn - what are your thoughts here? Elsewhere we try to limit to recommending a singular best model for a given provider, but this is an interesting case. Same here: https://github.com/continuedev/continue/pull/2632/files#diff-4f5610324759b3a04261998c57a2ae0bd117ca53d8b0af29b2838e5dd20a1f0eR61

Comment (TyDunn) [reply docs/docs/customize/model-providers/top-level/vertexai.md]: @Patrick-Erichsen To best help new users, I think we should keep it to our top recommendation here too, which would be `Claude 3.5 Sonnet` in this case

Comment (Lash-L) [reply docs/docs/customize/model-providers/top-level/vertexai.md]: Changed!

**Comment by Lash-L** [comment]
I've made some changes to reduce the amount of code that copies over. But some has to be overridden, so I would have had to make a bunch of helper functions, so I decided to skip that for now

**Comment by Lash-L** [comment]
can likely close #1573 as well

**Comment by Lash-L** [comment]
Hey @Patrick-Erichsen anything I can do to help this get merged? I've had to do a couple of large merge conflicts for dev, and I'd like to avoid that to make sure I don't accidentally make any breaking changes during dev merges

**Comment by Patrick-Erichsen** [comment]
Thanks for your patience here @Lash-L , and also for merging everything into the single provider class, looks great!

**Comment by Lash-L** [comment]
> Thanks for your patience here @Lash-L , and also for merging everything into the single provider class, looks great!

Thanks @Patrick-Erichsen how long is a typical release cycle? When should I expect to be able to use this in a production instance?

**Comment by Patrick-Erichsen** [comment]
We do our main releases on Fridays so it will be available shortly!

cc @sestinj , mind following up here after we release?

**Comment by sestinj** [comment]
Now released in pre-release, soon main

**Pull request merged**